core/drivers/bnxt/bnxt_fw.c

```diff
@@ -0,0 +1,684 @@
+					 BNXT_BSPD_CFG_OFFSET)
+#define BNXT_CONFIG_NS3Z_BSPD_DEST	(BNXT_CONFIG_NS3Z_DEST + \
+					 BNXT_BSPD_CFG_OFFSET)
+#define BNXT_BSPD_CFG_SIZE	0x200
+
+#define BNXT_CRASH_DUMP_INFO_NS3_BASE	0x3a5ff00
+#define BNXT_CRASH_DUMP_INFO_NS3Z_BASE	0x31fff00
+
+#define IS_ALIGNED(addr, algn)      (!((addr) & ((algn) - 1)))
+
+#define SZ_1K				0x400
+
+#define BUFFER_PADDING			SZ_1K
+
+#define INC_SRC_ADDR			1
+
+#define EOF				-1
+
+#define BCM_BNXT_FASTBOOT_MASK		0x3u
+#define BCM_BNXT_FASTBOOT_TYPE_1	1
+
+#define ADDR_IS_4BYTE_ALIGNED(addr)	IS_ALIGNED(addr, 4)
+
+#define SECTION_IS_LOADABLE(section_ptr) \
+	((section_ptr)->flags_src_offset & SECTION_FLAGS_IS_LOADABLE)
+#define SECTION_IS_ZIPPED(section_ptr) \
+	((section_ptr)->flags_src_offset & SECTION_FLAGS_IS_ZIPPED)
+#define SECTION_IS_TOBE_COPIED(section_ptr) \
+	((section_ptr)->flags_src_offset & \
+	 (SECTION_FLAGS_IS_EXEC_INSTR | SECTION_FLAGS_IS_DATA))
+#define SECTION_IS_TOBE_ZEROED(section_ptr) \
+	((section_ptr)->flags_src_offset & SECTION_FLAGS_IS_BSS)
+#define SECTION_IS_4BYTE_ALIGNED(section_ptr) \
+	ADDR_IS_4BYTE_ALIGNED((section_ptr)->dest_addr)
+
+#define SECTION_SRC_OFFSET(section_ptr) \
+	((section_ptr)->flags_src_offset & SECTION_SRC_OFFFSET_MASK)
+
+/* -------------------------------------------------------------------------- */
+
+/* Section header for each image block */
+struct ape_section_hdr_s {
+	/* destination address that this section is to be copied to */
+	uint32_t dest_addr;
+
+	/*
+	 * bit[0:23]  source offset address that this image copy from
+	 * bit[24:31] flags
+	 */
+	uint32_t flags_src_offset;
+#define SECTION_FLAGS_MASK		0xff000000
+	/* session is compressed (zipped) */
+#define SECTION_FLAGS_IS_ZIPPED		0x01000000
+	/* session contains CRC */
+#define SECTION_FLAGS_IS_CRC		0x02000000
+	/* session contains executable code (e.g. .text) */
+#define SECTION_FLAGS_IS_EXEC_INSTR	0x04000000
+	/* session contains initialized data (e.g. .data) */
+#define SECTION_FLAGS_IS_DATA		0x08000000
+	/* session contains zero initialized data (e.g. .bss) */
+#define SECTION_FLAGS_IS_BSS		0x10000000
+	/* loadable section mask */
+#define SECTION_FLAGS_IS_LOADABLE	(SECTION_FLAGS_IS_EXEC_INSTR | \
+					 SECTION_FLAGS_IS_DATA | \
+					 SECTION_FLAGS_IS_BSS)
+#define SECTION_SRC_OFFFSET_MASK	0x00ffffff
+
+	/* original image length, dword (4byte) length */
+	uint32_t org_data_len;
+
+	/* compressed image length (if FlAGS_IS_ZIPPED is set) */
+	uint32_t zip_data_len;
+
+	/*
+	 * checksum value for this image block, if FLAGS_IS_CRC then
+	 * this is CRC checksum; otherwise it is a simple summation
+	 */
+	uint32_t checksum;
+};
+
+struct version_s {
+	uint8_t version[16];	/* Null-terminated file version string */
+};
+
+struct ver_ext_offset_s {
+	uint8_t  version[12];	/* Null-terminated file version string */
+	uint32_t ext_hdr_offset;
+};
+
+union version_and_offset_u {
+	struct version_s version1;
+	struct ver_ext_offset_s	version2;
+};
+
+#pragma pack(1)
```

Comment:
I don't know what this means, but I suppose it has something to do with the __packed attribute which we have. Could that be used where needed instead?

Reply:
@shitalt 
#pragma pack(1)
  xxx
#pragma pack()
can be replaced by 
 struct ape_bin_hdr_s {
 xxx
} __packed;
 or __packed  __aligned(1);    but for sure test the output.